akka-javasdk/src/main/scala/akka/javasdk/impl/agent/ChatAgentImpl.scala

```diff
@@ -0,0 +1,142 @@
+      override val metadata: Metadata,
+      span: Option[Span],
+      tracerFactory: () => Tracer)
+      extends AbstractContext
+      with ChatAgentContext {
+    override def tracing(): Tracing = new SpanTracingImpl(span, tracerFactory)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[impl] final class ChatAgentImpl[A <: ChatAgent](
+    componentId: String,
+    sessionId: Optional[String],
+    val factory: ChatAgentContext => A,
+    tracerFactory: () => Tracer,
+    serializer: JsonSerializer,
+    componentDescriptor: ComponentDescriptor,
+    regionInfo: RegionInfo)
+    extends SpiChatAgent {
+  import ChatAgentImpl._
+
+  private val traceInstrumentation = new TraceInstrumentation(componentId, ChatAgentCategory, tracerFactory)
+
+  private val router: ReflectiveChatAgentRouter = {
+    val agentContext = new ChatAgentContextImpl(sessionId, regionInfo.selfRegion, Metadata.EMPTY, None, tracerFactory)
+    new ReflectiveChatAgentRouter(factory(agentContext), componentDescriptor.methodInvokers, serializer)
+  }
+
+  override def handleCommand(command: SpiAgent.Command): Future[SpiChatAgent.Effect] = {
+
+    val span: Option[Span] =
+      traceInstrumentation.buildSpan(ComponentType.ChatAgent, componentId, None, command.metadata)
+    span.foreach(s => MDC.put(Telemetry.TRACE_ID, s.getSpanContext.getTraceId))
+    // smuggling 0 arity method called from component client through here
+    val cmdPayload = command.payload.getOrElse(BytesPayload.empty)
+    val metadata: Metadata = MetadataImpl.of(command.metadata)
+    val agentContext = new ChatAgentContextImpl(sessionId, regionInfo.selfRegion, metadata, span, tracerFactory)
+
+    try {
+      val commandEffect = router
+        .handleCommand(command.name, cmdPayload, agentContext)
+        .asInstanceOf[ChatAgentEffectImpl[AnyRef]] // FIXME improve?
+
+      def errorOrReply: Either[SpiAgent.Error, (BytesPayload, SpiMetadata)] = {
+        commandEffect.secondaryEffect match {
+          case ErrorReplyImpl(description) =>
+            Left(new SpiAgent.Error(description))
+          case MessageReplyImpl(message, m) =>
+            val replyPayload = serializer.toBytes(message)
+            val metadata = MetadataImpl.toSpi(m)
+            Right(replyPayload -> metadata)
+          case NoSecondaryEffectImpl =>
+            throw new IllegalStateException("Expected reply or error")
+        }
+      }
+
+      commandEffect.primaryEffect match {
+        case req: RequestModel =>
+          val metadata = MetadataImpl.toSpi(req.replyMetadata)
+          Future.successful(new SpiChatAgent.RequestModelEffect(req.systemMessage, req.userMessage, metadata))
+
+        case NoPrimaryEffect =>
+          errorOrReply match {
+            case Left(err) =>
+              Future.successful(new SpiChatAgent.ErrorEffect(err))
+            case Right((reply, metadata)) =>
+              Future.successful(new SpiChatAgent.ReplyEffect(reply, metadata))
+          }
+      }
```

Comment:
All branches do Future.successful so that could be pulled out and done in one place